Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,461 @@
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter : Prop where
+  symmetric : M.IsSymm := by aesop
+  diagonal : ∀ i : B, M i i = 1 := by aesop
+  off_diagonal : ∀ i j : B, i ≠ j → M i j ≠ 1 := by aesop
+
+namespace CoxeterGroup
+
+namespace Relations
+
+/-- The relations corresponding to a Coxeter matrix. -/
+def ofMatrix : B × B → FreeGroup B :=
+ Function.uncurry fun i j => (FreeGroup.of i * FreeGroup.of j) ^ M i j
+
+/-- The set of relations corresponding to a Coxeter matrix. -/
+def toSet : Set (FreeGroup B) :=
+  Set.range <| ofMatrix M
+
+end Relations
+
+end CoxeterGroup
+
+/-- The group presentation corresponding to a Coxeter matrix. -/
+def Matrix.CoxeterGroup := PresentedGroup <| CoxeterGroup.Relations.toSet M
+
+instance : Group (Matrix.CoxeterGroup M) := by
+  exact QuotientGroup.Quotient.group _
+
+namespace CoxeterGroup
+
+def of (b : B) : Matrix.CoxeterGroup M := PresentedGroup.of b
+
+@[simp]
+theorem of_apply (b : B) : of M b = PresentedGroup.of (rels := Relations.toSet M) b :=
+  rfl
+
+end CoxeterGroup
+
+/-- A Coxeter system `CoxeterSystem W` is a structure recording the isomorphism between
+a group `W` and the group presentation corresponding to a Coxeter matrix. Equivalently, this can
+be seen as a list of generators of `W` parameterized by the underlying type of `M`, which
+satisfy the relations of the Coxeter matrix `M`. -/
+structure CoxeterSystem (W : Type*) [Group W]  where
+  /-- `CoxeterSystem.ofMulEquiv` constructs a Coxeter system given an equivalence with the group
+  presentation corresponding to a Coxeter matrix `M`. -/
+  ofMulEquiv ::
+    /-- `mulEquiv` is the isomorphism between the group `W` and the group presentation
+    corresponding to a Coxeter matrix `M`. -/
+    mulEquiv : W ≃* Matrix.CoxeterGroup M
+
+/-- Coxeter system of matrix `M` on the group `M.CoxeterGroup`. -/
+def Matrix.coxeterSystemCoxeterGroup : CoxeterSystem M M.CoxeterGroup :=
+  CoxeterSystem.ofMulEquiv (MulEquiv.refl _)
+
+/-- A group is a Coxeter group if it is registered in a Coxeter system
+ for some Coxeter matrix `M`. -/
+class IsCoxeterGroup (W : Type u) [Group W] : Prop where
+  nonempty_system : ∃ (B : Type u), ∃ (M : Matrix B B ℕ),
+    M.IsCoxeter ∧ Nonempty (CoxeterSystem M W)
+
+namespace CoxeterSystem
+
+open Matrix
+
+variable {B B' W H : Type*} [Group W] [Group H]
+
+variable {M : Matrix B B ℕ} [M.IsCoxeter]
+
+/-- A Coxeter system for `W` with Coxeter matrix `M` indexed by `B`, is associated to
+a map `B → W` recording the images of the indices. -/
+instance funLike : FunLike (CoxeterSystem M W) B (fun _ => W) where
+  coe cs := fun b => cs.mulEquiv.symm (.of b)
+  coe_injective' := by
+    rintro ⟨cs⟩ ⟨cs'⟩ hcs'
+    have H : (cs.symm : CoxeterGroup M →* W) = (cs'.symm : CoxeterGroup M →* W) := by
+      unfold CoxeterGroup
+      ext i; exact congr_fun hcs' i
+    have : cs.symm = cs'.symm := by ext x; exact FunLike.congr_fun H x
```

Comment:
All the `.symm`s suggest flipping the direction of the `MulEquiv` in the definition above. But I have not yet read the rest of the file, so I don't know what the optimal direction is.

Reply:
This is a good question.  For context, I followed the direction of `FreeGroupBasis` in `Mathlib.GroupTheory.FreeGroup.IsFreeGroup`.

Reply:
Ok, let's keep it like this for now. But I can imagine we might want to flip later (maybe in both places).